### PR TITLE
Cross Cluster Search: propagate original indices per cluster

### PR DESCRIPTION
**Pull request by javanna**
In case of a Cross Cluster Search, the coordinating node should split the original indices per cluster, and send over to each cluster only its own set of original indices, rather than set taken from the original search request which contains all the indices.

In fact, each remote cluster should not be aware of the indices belonging to other remote clusters.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -494,11 +496,12 @@ final SearchContext createContext(ShardSearchRequest request, @Nullable Engine.S
...
     }
 
-    public DefaultSearchContext createSearchContext(ShardSearchRequest request, TimeValue timeout, @Nullable Engine.Searcher searcher)
+    public DefaultSearchContext createSearchContext(ShardSearchRequest request, OriginalIndices originalIndices,
+                                                    TimeValue timeout, @Nullable Engine.Searcher searcher)
         throws IOException {
         IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
         IndexShard indexShard = indexService.getShard(request.shardId().getId());
-        SearchShardTarget shardTarget = new SearchShardTarget(clusterService.localNode().getId(), indexShard.shardId());
+        SearchShardTarget shardTarget = new SearchShardTarget(clusterService.localNode().getId(), indexShard.shardId(), originalIndices);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: I don't think we need this here at all lets pass always null here to the `SearchShardTarget`

File: core/src/main/java/org/elasticsearch/cluster/routing/GroupShardsIterator.java [hunk 2]
```diff
@@ -30,14 +30,14 @@
  * ShardsIterators are always returned in ascending order independently of their order at construction
  * time. The incoming iterators are sorted to ensure consistent iteration behavior across Nodes / JVMs.
 */
-public final class GroupShardsIterator implements Iterable<ShardIterator> {
+public final class GroupShardsIterator<S extends ShardIterator> implements Iterable<S> {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/GroupShardsIterator.java]: `s/S/ShardIt`

File: core/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java [hunk 3]
```diff
@@ -112,7 +112,7 @@ private void innerRun() throws IOException {
             final IntArrayList[] docIdsToLoad = searchPhaseController.fillDocIdsToLoad(numShards, reducedQueryPhase.scoreDocs);
             if (reducedQueryPhase.scoreDocs.length == 0) { // no docs to fetch -- sidestep everything and return
                 phaseResults.stream()
-                    .map(e -> e.queryResult())
+                    .map(SearchPhaseResult::queryResult)
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java]: 👍

File: core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 4]
```diff
@@ -268,27 +278,30 @@ private void executeSearch(SearchTask task, SearchTimeProvider timeProvider, Sea
...
+                                                             OriginalIndices locallIndices,
+                                                             List<SearchShardIterator> remoteShardIterators) {
+        List<SearchShardIterator> shards = new ArrayList<>();
+        for (SearchShardIterator shardIterator : remoteShardIterators) {
             shards.add(shardIterator);
         }
         for (ShardIterator shardIterator : localShardsIterator) {
-            shards.add(shardIterator);
+            List<ShardRouting> shardRoutings = new ArrayList<>();
+            for (ShardRouting shardRouting : shardIterator.asUnordered()) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: I hate this copying here :) can we maybe do this:
 * make `ShardsIterator extends Iterable<ShardRouting>`
 * remove `ShardsIterator#asUnordered()`
 * add `List<ShardRouting> ShardsIterator#getShardRoutings()`

that way we can just reuse the list?

**Review by s1monw — CHANGES_REQUESTED** [review]
i left some minors looks good man!

File: core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java (hunk 4 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: loving this plan. I don't like this copying either.

File: core/src/main/java/org/elasticsearch/search/SearchService.java (hunk 1 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/SearchService.java]: thanks for bringing this up, I was not too sure about this.

**Comment by javanna** [comment]
@s1monw I addressed your comments and added some tests. Can you have another look please?

File: core/src/main/java/org/elasticsearch/cluster/routing/PlainShardsIterator.java [hunk 5]
```diff
@@ -74,7 +75,12 @@ public int sizeActive() {
     }
 
     @Override
-    public Iterable<ShardRouting> asUnordered() {
+    public List<ShardRouting> getShardRoutings() {
         return shards;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/PlainShardsIterator.java]: Maybe wrap it in an unmodifiable list in the ctor? I think we should not allow mutation here?

**Review by s1monw — APPROVED** [review]
left one suggestion LGTM otherwise. no need for another review. Thanks for the additional tests

File: core/src/main/java/org/elasticsearch/cluster/routing/PlainShardsIterator.java (hunk 5 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/routing/PlainShardsIterator.java]: ++ will do

**Pull request merged**